iocore/net/TLSSNISupport.cc

```diff
@@ -64,8 +67,12 @@ TLSSNISupport::perform_sni_action()
   }
 
   SNIConfig::scoped_config params;
-  if (const auto &actions = params->get({servername, std::strlen(servername)}); !actions.first) {
-    Debug("ssl_sni", "%s not available in the map", servername);
+  // should always work in this context of SSL action callbacks
+  SSLNetVConnection *ssl_vc{dynamic_cast<SSLNetVConnection *>(this)};
```

Comment:
I took a look this, and I think it might be time to refactor. 

```
class SSLNetVConnection : public UnixNetVConnection,
                          public ALPNSupport,
                          public TLSSessionResumptionSupport,
                          public TLSSNISupport,
                          public TLSEarlyDataSupport,
                          public TLSTunnelSupport,
                          public TLSCertSwitchSupport,
                          public TLSBasicSupport
```

This is where we obtain the TLSSNIConfig:
```cpp
TLSSNISupport *
TLSSNISupport::getInstance(SSL *ssl)
{
  return static_cast<TLSSNISupport *>(SSL_get_ex_data(ssl, _ex_data_index));
}
```

and SSLUtils.cc does the actual calls to perform SNI actions, on the `TLSSNISupport` object it got from this... apparently some global data stored using OpenSSL? Am I reading that right?

So there's no NetVConnection in sight. The only way we can obtain it is to `dynamic_cast`.  I'm not the only person who's noticed this:

```cpp
    // TODO: change design to avoid this dynamic_cast
    auto ssl_vc = dynamic_cast<SSLNetVConnection *>(snis);
```
    
@masaori335 put in this TODO, maybe he's thought some about how to do this. I suggest we aim to do the design change in a separate PR, as it will probably be nontrivial.

Reply:
No, the goal is not to use `SSLNetVConnection` directly. You should not try to get the instance in any way. This doesn't work if the connection is QUIC.

Reply:
Ahh, yes, this looks like a mistake in my change. We need to dynamic_cast to an `I_NetVConnection*` I think.

Reply:
Or are you saying we can't get the connection at all? We have to find the incoming port in order to filter on it.

Reply:
There are lots of other places where it is explicitly dynamic_cast to an `SSLNetVConnection*`.

Reply:
The existing dynamic_casts were written before QUIC, and we need to fix them. New code should try not to introduce a new one.

Here are two options but there may be better ways.
- Get a port number from SSL object
- Add a protected virtual function, which returns a port number, to SNISupport

Reply:
I added a base class inbetween the NetVConnection implementations and the UnixNetVConnection base class. This is probably not the cleanest solution, but it is easy, and I am thinking it would be better to do more extensive design changes separately. Will this work?

Reply:
It doesn't look like a right direction to me. What design change are you planning?

With this change for port number filtering, although port number is not directly related to SNI, port number becomes essential to support SNI (or more specifically, SNIAction) on a `NetVConnectoin` implementation. I'd require a `NetVConnection` implementation to have the capability of providing the local port number by adding `virtual int _get_local_port() = 0` to `TLSSNISupport`. If one cannot fulfill the requirement, the implementation can't say it supports "SNI".

Reply:
Do you have an example of where we need to use SNI and do not want the `NetVConnection` to have a local port number? If that's because there is none, that implies we need to re-evaluate this whole idea of filtering by port.

Reply:
It's fine to have a local port number. I'm just talking about how to access it.

`NetVConnection` can keep current `get_local_port()`, and I'm suggesting a protected new function `_get_local_prot()` so `TLSSNISupport` can access it without depending on any `NetVConnection`. The new function is an interface between `TLSSNISupport` and something that supports `TLSSNISupport`. `TLSSNISupport` itself doesn't need `NetVConnection` (i.e. `TLSSNISupport` doesn't need to transfer data over network).

Reply:
This is what I'm suggesting. You can simply call `this->_get_local_port()` here without dynamic_cast.
https://github.com/apache/trafficserver/compare/master...maskit:export_local_port?expand=1

Reply:
Ah, very good! I was looking for how to move `get_local_port()` into `TLSSNISupport`, but it did not occur to me to simply use a different name! I like that approach, so let's go with that.